man/functools.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "FUNCTOOLS" "1" "October 2011" "" ""
.
.SH "NAME"
\fBfunctools\fR \- a minimal javascript library for functional programming
.
.SH "SYNOPSIS"
A typical usage:
.
.IP "" 4
.
.nf

var map = require("functools")\.map;

var seq = [3,1,4,1,5,9];

functools\.map(function(el,ind,seq){
  return el+ind;
},seq);
.
.fi
.
.IP "" 0
.
.P
Function Composition:
.
.IP "" 4
.
.nf

var compose = require("functools")\.compose;

compose(select, update, prettify, display)("body \.messages");
.
.fi
.
.IP "" 0
.
.P
Currying:
.
.IP "" 4
.
.nf

var fn = require("functools");

var pickEvens = fn\.curry(fn\.filter)(function(num){ return num%2==0 });

pickEvens([3,1,4]) // returns [4]
pickEvens([1,5,9,2,6,5]) // returns [2,6]
.
.fi
.
.IP "" 0
.
.P
Partial Function Application:
.
.IP "" 4
.
.nf

Foobar\.prototype\.doSomething = function(msg){
  puts(msg);
  this\.corge();
}

server\.on("connection", functools\.partial(foobar\.doSomething, ["Hello World"], foobar));
.
.fi
.
.IP "" 0
.
.SH "DESCRIPTION"
functools is a library for functional programming written in JavaScript\. It\'s based on a CommonJS module consists of several function manipulation and list iteration tools\.
.
.SH "API"
.
.SS "compose(<em>functions \.\.\.</em>)(<em>value</em>)"
Combine \fIfunctions\fR in a new one, passing the result of each function to next one, from left to right\.
.
.IP "" 4
.
.nf

function cube(x){ return x*x*x };

compose(Math\.sqrt,cube)(4); // returns 8
.
.fi
.
.IP "" 0
.
.SS "compose\.async(<em>functions \.\.\.</em>)(<em>value</em>,<em>callback</em>)"
Asynchronous, continuation passing based version of compose function\. Requires specified functions to call a callback function, passing an error object (if there is) and the result to be carried\.
.
.IP "" 4
.
.nf

function cube(x,callback){ callback(undefined, x*x*x); }

function sqrt(x,callback){ callback(undefined, Math\.sqrt(x)); }

compose\.async(Math\.sqrt, cube)(4,function(error, result){
  console\.log(error); // undefined
  console\.log(result); // puts "8"
});
.
.fi
.
.IP "" 0
.
.SS "curry(<em>function</em>, <em>args \.\.\.</em>)"
Transform multiple\-argument \fIfunction\fR into a chain of functions that return each other until all arguments are gathered\.
.
.IP "" 4
.
.nf

function sum(x,y){ return x+y; }

var add3 = curry(sum, 3);

add3(14); // returns 17
add3(20); // returns 23
.
.fi
.
.IP "" 0
.
.SS "partial(<em>function</em>,<em>initial arguments</em>,<em>context</em>)"
Return a new function which will call \fIfunction\fR with the gathered arguments\.
.
.IP "" 4
.
.nf

function testPartial(){
  var args = reduce(function(x,y){ x+", "+y },arguments);

  console\.log("this:",this);
  console\.log("args:",args);
}

partial(testPartial, [3,14], 3\.14159)(1,5,9);
.
.fi
.
.IP "" 0
.
.P
The example code above will output:
.
.IP "" 4
.
.nf

this: 3\.14159
args: 3,14,1,5,9
.
.fi
.
.IP "" 0
.
.SS "each(<em>function</em>,<em>iterable</em>)"
Call \fIfunction\fR once for element in \fIiterable\fR\.
.
.IP "" 4
.
.nf

each(function(el,ind,list){ console\.assert( el == list[ind] ); }, [3,1,4]);
.
.fi
.
.IP "" 0
.
.SS "map(<em>function</em>,<em>iterable</em>)"
Invoke \fIfunction\fR once for each element of \fIiterable\fR\. Creates a new array containing the values returned by the function\.
.
.IP "" 4
.
.nf

map(function(el,ind,list){ return el*el },[3,1,4,1,5,9]); // returns [9,1,16,1,25,81]
.
.fi
.
.IP "" 0
.
.SS "map\.async(<em>function</em>,<em>iterable</em>, <em>callback</em>)"
Apply async \fIfunction\fR to every item of \fIiterable\fR, receiving a callback function which takes error (if there is) and replacement parameters\.
.
.IP "" 4
.
.nf

map(function(el,callback){ callback(undefined,el*el); },[3,1,4,1,5,9],function(error, newArray){
  assert\.equal(newArray[0], 9); // true
});
.
.fi
.
.IP "" 0
.
.SS "filter(<em>function</em>,<em>iterable</em>)"
Construct a new array from those elements of \fIiterable\fR for which \fIfunction\fR returns true\.
.
.IP "" 4
.
.nf

filter(function(el,ind,list){ return el%2==0 },[3,1,4]); // returns [4]
.
.fi
.
.IP "" 0
.
.SS "filter\.async(<em>function</em>,<em>iterable</em>, <em>callback</em>)"
Call async \fIfunction\fR once for each element in \fIiterable\fR, receiving a boolean parameter, and construct a new array of all the values for which \fIfunction\fR produces \fItrue\fR
.
.IP "" 4
.
.nf

filter(function(el,callback){ callback(el%2==0); },[3,1,4],function(newArray){
  assert\.equal(newArray[0], 4); // true
});
.
.fi
.
.IP "" 0
.
.SH "juxt(<em>functions \.\.\.</em>)"
Take a set of functions, return a function that is the juxtaposition of those functions\. The returned function takes a variable number of arguments and returns a list containing the result of applying each fn to the arguments\.
.
.IP "" 4
.
.nf

function inc1(n){ return n+1 };
function inc2(n){ return n+2 };
function inc3(n){ return n+3 };

juxt(inc1, inc2, inc3)([3,1,4]); // returns [4,3,7]
.
.fi
.
.IP "" 0
.
.SH "juxt\.async(<em>functions \.\.\.</em>)"
Async implementation of \fIjuxt\fR\.
.
.IP "" 4
.
.nf

function md5(path, callback){ fetch(path, callback); }
function sha1(path, callback){ fetch(path, callback); }
function crc32(path, callback){ fetch(path, callback); }

juxt\.async(md5, sha1, crc32)("hello world", function(error,  result){
  result[0] => md5("hello world")
  result[1] => sha1("hello world")
  result[2] => crc32("hello world")
});
.
.fi
.
.IP "" 0
.
.SS "reduce(<em>function</em>,<em>iterable</em>)"
Apply \fIfunction\fR cumulatively to the items of \fIiterable\fR, as to reduce the \fIiterable\fR to a single value
.
.IP "" 4
.
.nf

reduce(function(x,y){ return x*y }, [3,1,4]); // returns 12
.
.fi
.
.IP "" 0
.
.SS "reduce\.async(<em>function</em>,<em>iterable</em>, <em>callback</em>)"
Apply version of \fIreduce\fR\. See the example below\.
.
.IP "" 4
.
.nf

reduce(function(x,y,callback){ callback(undefined,x*y); },[3,1,4],function(error, result){
  assert\.equal(result, 12); // true
});
.
.fi
.
.IP "" 0
.
.SH "SEE ALSO"
.
.IP "\(bu" 4
Functional Programming \- Eloquent JavaScript \fIhttp://eloquentjavascript\.net/chapter6\.html\fR
.
.IP "\(bu" 4
Underscore\.js \fIhttp://documentcloud\.github\.com/underscore/\fR
.
.IP "\(bu" 4
Roka Lisp \fIhttp://github\.com/azer/roka\fR
.
.IP "" 0



lib/functools.js
var functools = (function(undefined){

  /**
   * Function composition implementation
   */ 
  function compose(/* functions */){
    var fns = Array.prototype.slice.call(arguments);
    return function(initialValue){
      return reduce(function(f,g){
        return function(val){ 
          return g(f(val));
        }
      },fns)(initialValue);
    }
  }

  compose.async = function compose_async(/* functions */){
    var fns = Array.prototype.slice.call(arguments);
    return function(initialValue,callback){
      (function(i,error,value){
        if(error || fns.length<=i){
          return callback(error, value);
        }

        fns[i](value, partial(arguments.callee, [i+1]));

      })(0,undefined,initialValue);
    }
  }

  /**
   * Take a set of functions, return a function that is the juxtaposition of those
   * functions. The returned function takes a variable number of arguments and returns
   * a list containing the result of applying each fn to the arguments.
   */
  function juxt(/* functions */){
    var fns = arguments;
    return function(){
      var args = arguments;
      return map(function(fn){
        return fn.apply(undefined, args); 
      }, fns);    
    }
  }

  juxt.async = function(/* functions */){
    var fns = arguments;
    return function(/* args,  */callback){
      var args = Array.prototype.slice.call(arguments, 0, arguments.length-1),
          callback = arguments[arguments.length-1];

      map.async(function(fn, callback){
        fn.apply(undefined, args.concat([callback]));
      }, fns, callback);
    }
  }

  /**
   * Transform multiple-argument 'fn' into a chain of functions that return each other until all arguments
   * are gathered.
   */
  function curry(fn){
    var args = Array.prototype.slice.call(arguments,1),
        len = fn.length;
    return (function(){
      Array.prototype.push.apply(args,arguments);
      return args.length >= len && fn.apply(null,args) || arguments.callee;
    })();
  }

  /**
   * Execute 'fn' once per 'iterable' element.
   */
  function each(fn,iterable){
    for(var i = -1, len=iterable.length; ++i < len; ){
      fn(iterable[i],i,iterable);
    };
    return iterable;
  }

  /**
   * Apply 'fn' to every element of 'iterable', returning those elements for which fn returned a true value.
   */
  function filter(fn,iterable){
    var accumulation = [];
    for(var i = -1, len=iterable.length; ++i < len; ){
      fn(iterable[i],i,iterable) && accumulation.push(iterable[i]);
    };
    return accumulation;
  }

  filter.async = function filter_async(fn, iterable, callback){
    var accumulation = [];
    (function(i,ptest){

        ptest && accumulation.push(iterable[i-1]);

        if(i>=iterable.length){
          return callback(accumulation);
        }

        fn(iterable[i], partial(arguments.callee, [i+1]));

    })(0);
  }

  /**
   * Apply fn to every element of iterable.
   */
  function map(fn,iterable){
    var seq = Array.prototype.slice.call(iterable);
    for(var i = -1, len=seq.length; ++i < len; ){
      seq[i] = fn(seq[i],i,seq);
    };

    return seq;
  }

  map.async = function map_async(fn, iterable, callback){
    var seq = Array.prototype.slice.call(iterable);
    (function(i, error, rpl){
      
      rpl && ( seq[i-1] = rpl );

      if(error || i>=seq.length){
        return callback(error, seq);
      }

      fn(seq[i], partial(arguments.callee, [i+1]));

    })(0);
  }

  /**
   * Return a new function which will call 'fn' with the positional arguments 'args'
   */
  function partial(fn,initialArgs,ctx){
    !initialArgs && ( initialArgs = [] );
    return function(){
      var args = Array.prototype.slice.call(initialArgs,0);
      Array.prototype.push.apply(args,arguments);
      return fn.apply(ctx,args);
    };
  };

  /**
   * Apply fn cumulatively to the items of iterable,  as to reduce the iterable to a single value
   */
  function reduce(fn,iterable){
    var accumulator = iterable[0];

    for(var i = 0, len=iterable.length; ++i < len; ){
      accumulator=fn(accumulator,iterable[i],iterable);
    };

    return accumulator;
  }

  reduce.async = function reduce_async(fn,iterable,callback){
    (function(i, error, accumulator){
      
      if(error || i>=iterable.length){
        return callback(error, accumulator);
      }

      fn(accumulator, iterable[i], partial(arguments.callee, [i+1]));

    })(1,undefined,iterable[0]);
  }

  return {
    "compose":compose,
    "curry":curry,
    "each":each,
    "filter":filter,
    'juxt':juxt,
    "map":map,
    "partial":partial,
    "reduce":reduce
  }

})();

if(typeof module != 'undefined' && module.exports){
  module.exports = functools;
}
